### ISSUE TITLE
Customization of BlockDiagonalMask or BlockDiagonalCausalMask

### ISSUE THREAD

**Issue by KJ-rc** [facebookresearch/xformers#990]
# ❓ Questions and Help
I am trying to write a customization of BlockDiagonalMask with an additional variable `group_lengths_list`. 
Inside each block, tokens are separated into 3 groups (number of groups is a hyper-param), all tokens can see the tokens within the same group and those in the last group (the 3rd group here). For example, 2 blocks on the diagonal with `group_lengths_list: [1, 1, 2] and [2, 2, 3]` would be like
```
1 0 1 1   0 0 0 0 0 0 0 
0 1 1 1   0 0 0 0 0 0 0 
0 0 1 1   0 0 0 0 0 0 0
0 0 1 1   0 0 0 0 0 0 0 

0 0 0 0   1 1 0 0 1 1 1
0 0 0 0   1 1 0 0 1 1 1
0 0 0 0   0 0 1 1 1 1 1
0 0 0 0   0 0 1 1 1 1 1
0 0 0 0   0 0 0 0 1 1 1
0 0 0 0   0 0 0 0 1 1 1
0 0 0 0   0 0 0 0 1 1 1
```
I was trying to realize it by making a customization of `BlockDiagonalMask` or `BlockDiagonalCausalMask`, like `BlockDiagonalCausalLocalAttentionMask`, but I cannot pass `group_lengths` into `_create_block_mask` (or, identifying current block, e.g., block index).

Is there any workaround?

**Comment by danthe3rd** [comment]
Hi,
Which hardware are you running on? Are you interested in the BW pass as well?
If you have an Sm80+ device with Flash-Attention available, I would look at the problem as the addition of 2 attentions: one with mask A (BlockDiagonal), and one with mask B (sort of BlockDiagonal) - however I'm not sure if everything would work correctly to run this within xFormers.
Another way would be to materialize the full attention mask - but this will take memory and be quite slower obviously...

------------

**Mask A:**
```
1 0 0 0   0 0 0 0 0 0 0 
0 1 0 0   0 0 0 0 0 0 0 
0 0 1 1   0 0 0 0 0 0 0
0 0 1 1   0 0 0 0 0 0 0 

0 0 0 0   1 1 0 0 0 0 0
0 0 0 0   1 1 0 0 0 0 0
0 0 0 0   0 0 1 1 0 0 0
0 0 0 0   0 0 1 1 0 0 0
0 0 0 0   0 0 0 0 1 1 1
0 0 0 0   0 0 0 0 1 1 1
0 0 0 0   0 0 0 0 1 1 1
```
**Mask B:**
```
0 0 1 1   0 0 0 0 0 0 0 
0 0 1 1   0 0 0 0 0 0 0 
0 0 0 0   0 0 0 0 0 0 0
0 0 0 0   0 0 0 0 0 0 0 

0 0 0 0   0 0 0 0 1 1 1
0 0 0 0   0 0 0 0 1 1 1
0 0 0 0   0 0 0 0 1 1 1
0 0 0 0   0 0 0 0 1 1 1
0 0 0 0   0 0 0 0 0 0 0 
0 0 0 0   0 0 0 0 0 0 0 
0 0 0 0   0 0 0 0 0 0 0 
```

**Comment by KJ-rc** [comment]
Hi, @danthe3rd 

Thank you for the response. 
My device is A6000, and I'm also interested in backward.

I took some time to go through Flash-Attention (and Triton) and your proposal.
I found another wanted feature I missed, that `group_lengths` varies every forward pass.

- In Flash-Attention, they provide `FlashBlocksparseAttention` but "BlockMasks" is given and fixed after AttnLayer is initialized. I need to start with a lower-level function `flash_blocksparse_attn_func`.
- In Triton, I am still reading their tutorial for block pointers to see whether it is simpler.

Do you think they are in the right direction? 
If you have some suggestions or hints, I really appreciate them.

**Comment by FuyuanChen** [comment]
So, what method now？ 🤔
